{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GeneID      113219467     57801      9636  100288175  102465434  109623456  \\\n",
      "GSM6745599  -0.158632  1.070933  1.152150   0.499194  -0.342997  -0.697982   \n",
      "GSM6745600   1.379276  0.767235  0.761592  -0.765431  -0.342997  -0.697982   \n",
      "GSM6745601   5.355088  3.457126  2.062135   1.706335  -0.342997  -0.697982   \n",
      "GSM6745602   2.002351  0.897391  0.336385  -0.075635  -0.342997   0.697982   \n",
      "GSM6745603   0.679553  1.070933  0.673147   0.614160  -0.342997   0.697982   \n",
      "GSM6745604  -0.213027  1.548171  0.828458  -0.018153  -0.342997  -0.697982   \n",
      "GSM6745605   0.009500  0.810621 -0.304008   1.016540  -0.342997  -0.697982   \n",
      "GSM6745606  -0.393521 -0.274013 -0.724960  -0.650465  -0.342997  -0.697982   \n",
      "GSM6745607  -0.274840  0.289997 -0.605817   0.384228  -0.342997  -0.697982   \n",
      "GSM6745609  -0.334181  1.070933 -0.549893   1.591370  -0.342997  -0.697982   \n",
      "GSM6745610  -0.339126  2.329107 -0.600042   0.211779  -0.342997   0.697982   \n",
      "GSM6745611  -0.299565 -0.577710  0.256450  -0.305567  -0.342997  -0.697982   \n",
      "GSM6745612  -0.344071 -0.360783  0.478324   0.326745  -0.342997  -0.697982   \n",
      "GSM6745613  -0.173467  0.463538  2.571836   1.936267  -0.342997  -0.697982   \n",
      "GSM6745614  -0.321818 -0.447554 -0.288507   0.211779  -0.342997   0.697982   \n",
      "GSM6745615  -0.035006  0.073070  0.994711   2.223682   2.915476   2.093947   \n",
      "GSM6745616  -0.306983 -0.621095 -0.180306   0.614160  -0.342997   0.697982   \n",
      "GSM6745617  -0.304510 -0.447554 -0.598826   0.384228  -0.342997  -0.697982   \n",
      "GSM6745618  -0.158632 -0.794637 -0.571472   0.844091  -0.342997  -0.697982   \n",
      "GSM6745619  -0.410829 -0.968178 -0.901851  -0.765431  -0.342997  -0.697982   \n",
      "GSM6745620  -0.358906 -0.838022 -0.748059  -0.765431  -0.342997  -0.697982   \n",
      "GSM6745621  -0.336653 -0.534325 -0.706724  -0.592982  -0.342997  -0.697982   \n",
      "GSM6745622  -0.264950 -0.360783 -0.581198  -0.248084  -0.342997  -0.697982   \n",
      "GSM6745623  -0.282258 -0.707866 -0.499743  -0.305567  -0.342997   2.093947   \n",
      "GSM6745624  -0.260005 -0.621095 -0.443819  -0.190601  -0.342997   0.697982   \n",
      "GSM6745625  -0.252587 -0.447554 -0.531656   2.166199  -0.342997  -0.697982   \n",
      "GSM6745626  -0.173467  1.418015  3.757187   0.269262  -0.342997  -0.697982   \n",
      "GSM6745627  -0.452862 -0.968178 -0.799728  -1.397743  -0.342997   0.697982   \n",
      "GSM6745628  -0.264950 -0.707866  0.269823  -0.018153  -0.342997   0.697982   \n",
      "GSM6745629  -0.324291 -0.404169 -0.448682  -0.592982  -0.342997  -0.697982   \n",
      "GSM6745630  -0.388576 -0.577710 -0.379384  -0.248084  -0.342997  -0.697982   \n",
      "GSM6745631  -0.200665 -0.490939 -0.138667  -0.133118  -0.342997  -0.697982   \n",
      "GSM6745632  -0.195719 -0.794637 -0.593051  -1.512709  -0.342997   2.093947   \n",
      "GSM6745633  -0.190774 -0.751251 -0.522234  -1.340260  -0.342997   2.093947   \n",
      "GSM6745634  -0.346543 -1.011563 -0.619798  -1.570192  -0.342997  -0.697982   \n",
      "GSM6745635  -0.358906 -0.751251 -0.663869  -1.225294   2.915476   0.697982   \n",
      "GSM6745636  -0.361378 -0.534325 -0.503086  -1.110328   2.915476   2.093947   \n",
      "GSM6745637  -0.344071 -0.274013 -0.636818  -1.167811   2.915476  -0.697982   \n",
      "\n",
      "GeneID          8510  107984872  112268220  105378607  ...  105377240  \\\n",
      "GSM6745599  0.160668  -0.602311  -0.164399  -0.541682  ...  -0.222988   \n",
      "GSM6745600 -1.275890  -0.602311  -0.164399  -0.541682  ...  -0.222988   \n",
      "GSM6745601  2.435216  -0.602311  -0.164399  -0.541682  ...  -0.222988   \n",
      "GSM6745602  0.160668  -0.602311  -0.164399  -0.541682  ...  -0.222988   \n",
      "GSM6745603  0.579663   0.136006  -0.164399  -0.541682  ...  -0.222988   \n",
      "GSM6745604  0.938803   0.136006  -0.164399  -0.541682  ...  -0.222988   \n",
      "GSM6745605  2.016221   2.350955  -0.164399  -0.541682  ...  -0.222988   \n",
      "GSM6745606 -0.797037   0.136006  -0.164399  -0.541682  ...  -0.222988   \n",
      "GSM6745607  0.459950   2.350955  -0.164399   0.462412  ...  -0.222988   \n",
      "GSM6745609  0.639520  -0.602311  -0.164399  -0.541682  ...  -0.222988   \n",
      "GSM6745610  1.776794  -0.602311  -0.164399  -0.541682  ...  -0.222988   \n",
      "GSM6745611 -0.737181   0.136006  -0.164399  -0.039635  ...  -0.222988   \n",
      "GSM6745612 -0.138615  -0.602311  -0.164399   0.462412  ...  -0.222988   \n",
      "GSM6745613  2.076077  -0.602311  -0.164399   0.462412  ...  -0.222988   \n",
      "GSM6745614 -0.198472   0.136006  -0.164399  -0.541682  ...  -0.222988   \n",
      "GSM6745615  1.357799   0.874322  -0.164399   1.466505  ...   5.426047   \n",
      "GSM6745616  0.160668  -0.602311  -0.164399  -0.039635  ...  -0.222988   \n",
      "GSM6745617 -0.258328   3.827589  -0.164399   0.462412  ...  -0.222988   \n",
      "GSM6745618 -0.437898   0.874322  -0.164399  -0.039635  ...  -0.222988   \n",
      "GSM6745619 -1.395603  -0.602311  -0.164399  -0.541682  ...  -0.222988   \n",
      "GSM6745620 -0.856894  -0.602311  -0.164399  -0.541682  ...  -0.222988   \n",
      "GSM6745621 -0.378041   0.136006  -0.164399  -0.039635  ...  -0.222988   \n",
      "GSM6745622 -0.497754  -0.602311  -0.164399   0.462412  ...  -0.222988   \n",
      "GSM6745623 -0.138615   0.136006  -0.164399  -0.541682  ...   2.601530   \n",
      "GSM6745624  0.160668  -0.602311  -0.164399  -0.541682  ...  -0.222988   \n",
      "GSM6745625 -0.138615   0.136006  -0.164399  -0.541682  ...  -0.222988   \n",
      "GSM6745626  1.118372  -0.602311  -0.164399   0.964458  ...  -0.222988   \n",
      "GSM6745627 -1.515316   0.136006  -0.164399  -0.541682  ...  -0.222988   \n",
      "GSM6745628  1.118372   1.612639  -0.164399   0.462412  ...  -0.222988   \n",
      "GSM6745629 -0.797037  -0.602311  -0.164399  -0.039635  ...  -0.222988   \n",
      "GSM6745630 -1.335746  -0.602311  -0.164399  -0.541682  ...  -0.222988   \n",
      "GSM6745631 -0.078759  -0.602311  -0.164399  -0.541682  ...  -0.222988   \n",
      "GSM6745632 -0.378041  -0.602311  -0.164399  -0.541682  ...  -0.222988   \n",
      "GSM6745633 -0.318185  -0.602311  -0.164399   4.980833  ...  -0.222988   \n",
      "GSM6745634 -0.916750  -0.602311  -0.164399   1.466505  ...  -0.222988   \n",
      "GSM6745635 -0.916750  -0.602311  -0.164399  -0.541682  ...  -0.222988   \n",
      "GSM6745636 -0.797037  -0.602311  -0.164399   0.462412  ...  -0.222988   \n",
      "GSM6745637 -0.856894   0.136006   6.082763  -0.541682  ...  -0.222988   \n",
      "\n",
      "GeneID         84559    401634  105377243  107987359      4572      4512  \\\n",
      "GSM6745599 -0.164399 -0.164399  -0.222988  -0.164399 -0.135784 -0.237304   \n",
      "GSM6745600 -0.164399 -0.164399  -0.222988  -0.164399  0.567824  1.302546   \n",
      "GSM6745601 -0.164399 -0.164399  -0.222988  -0.164399  3.460437  5.055768   \n",
      "GSM6745602 -0.164399 -0.164399  -0.222988  -0.164399  1.427790  2.327118   \n",
      "GSM6745603 -0.164399 -0.164399  -0.222988  -0.164399  2.131398  0.856884   \n",
      "GSM6745604 -0.164399 -0.164399  -0.222988  -0.164399  1.975041 -0.052044   \n",
      "GSM6745605 -0.164399 -0.164399  -0.222988  -0.164399  1.349611 -0.139682   \n",
      "GSM6745606 -0.164399 -0.164399  -0.222988  -0.164399  0.098752 -0.558824   \n",
      "GSM6745607 -0.164399 -0.164399  -0.222988  -0.164399  0.724182 -0.258147   \n",
      "GSM6745609 -0.164399 -0.164399  -0.222988  -0.164399 -0.604856 -0.342890   \n",
      "GSM6745610 -0.164399 -0.164399  -0.222988  -0.164399  1.505969 -0.292707   \n",
      "GSM6745611 -0.164399 -0.164399  -0.222988  -0.164399 -0.839392 -0.394787   \n",
      "GSM6745612 -0.164399 -0.164399  -0.222988  -0.164399 -0.213963 -0.383851   \n",
      "GSM6745613 -0.164399 -0.164399  -0.222988  -0.164399 -0.292141 -0.307034   \n",
      "GSM6745614 -0.164399 -0.164399  -0.222988  -0.164399 -0.370320 -0.374021   \n",
      "GSM6745615 -0.164399 -0.164399   5.426047  -0.164399  0.802360  0.436256   \n",
      "GSM6745616 -0.164399 -0.164399  -0.222988  -0.164399 -0.761214 -0.228693   \n",
      "GSM6745617 -0.164399 -0.164399  -0.222988  -0.164399 -0.057605 -0.280552   \n",
      "GSM6745618 -0.164399 -0.164399  -0.222988  -0.164399  0.020573  0.022144   \n",
      "GSM6745619 -0.164399 -0.164399  -0.222988  -0.164399 -0.761214 -0.699579   \n",
      "GSM6745620 -0.164399 -0.164399  -0.222988  -0.164399 -0.917571 -0.372344   \n",
      "GSM6745621 -0.164399 -0.164399  -0.222988  -0.164399 -0.370320 -0.369182   \n",
      "GSM6745622 -0.164399 -0.164399  -0.222988  -0.164399 -0.683035 -0.283639   \n",
      "GSM6745623 -0.164399 -0.164399   2.601530  -0.164399 -0.292141 -0.269997   \n",
      "GSM6745624 -0.164399 -0.164399  -0.222988  -0.164399 -0.526678 -0.155724   \n",
      "GSM6745625 -0.164399 -0.164399  -0.222988  -0.164399 -0.761214 -0.253155   \n",
      "GSM6745626 -0.164399 -0.164399  -0.222988  -0.164399  0.020573  0.100067   \n",
      "GSM6745627 -0.164399 -0.164399  -0.222988  -0.164399 -0.995750 -0.778035   \n",
      "GSM6745628 -0.164399 -0.164399  -0.222988  -0.164399 -0.526678 -0.286306   \n",
      "GSM6745629 -0.164399 -0.164399  -0.222988  -0.164399 -0.917571 -0.433196   \n",
      "GSM6745630 -0.164399 -0.164399  -0.222988  -0.164399 -0.995750 -0.606873   \n",
      "GSM6745631 -0.164399 -0.164399  -0.222988  -0.164399  0.020573 -0.096739   \n",
      "GSM6745632 -0.164399 -0.164399  -0.222988  -0.164399 -0.057605 -0.143264   \n",
      "GSM6745633 -0.164399 -0.164399  -0.222988  -0.164399 -0.448499  0.174674   \n",
      "GSM6745634  6.082763  6.082763  -0.222988  -0.164399 -0.839392 -0.536076   \n",
      "GSM6745635 -0.164399 -0.164399  -0.222988  -0.164399 -0.683035 -0.518510   \n",
      "GSM6745636 -0.164399 -0.164399  -0.222988  -0.164399 -0.448499 -0.269883   \n",
      "GSM6745637 -0.164399 -0.164399  -0.222988   6.082763 -0.604856 -0.352416   \n",
      "\n",
      "GeneID          4574  Edited (1) or Unedited (0)  editing mechanism  \n",
      "GSM6745599 -0.162382                           1                BE4  \n",
      "GSM6745600  1.353526                           1                BE4  \n",
      "GSM6745601  5.391090                           1                BE4  \n",
      "GSM6745602  1.941426                           1               ABE8  \n",
      "GSM6745603  0.658735                           1               ABE8  \n",
      "GSM6745604 -0.206110                           1               ABE8  \n",
      "GSM6745605 -0.016622                           1               Cas9  \n",
      "GSM6745606 -0.381023                           1               Cas9  \n",
      "GSM6745607 -0.269273                           1               Cas9  \n",
      "GSM6745609 -0.322719                           1          Utelectro  \n",
      "GSM6745610 -0.337295                           1          Utelectro  \n",
      "GSM6745611 -0.293567                           1                BE4  \n",
      "GSM6745612 -0.337295                           1                BE4  \n",
      "GSM6745613 -0.172100                           1                BE4  \n",
      "GSM6745614 -0.317860                           1               ABE8  \n",
      "GSM6745615 -0.016622                           1               ABE8  \n",
      "GSM6745616 -0.303284                           1               ABE8  \n",
      "GSM6745617 -0.308143                           1               Cas9  \n",
      "GSM6745618 -0.152665                           1               Cas9  \n",
      "GSM6745619 -0.410175                           1               Cas9  \n",
      "GSM6745620 -0.361588                           1          Utelectro  \n",
      "GSM6745621 -0.332436                           1          Utelectro  \n",
      "GSM6745622 -0.259556                           1          Utelectro  \n",
      "GSM6745623 -0.264415                           1              dCas9  \n",
      "GSM6745624 -0.259556                           1              dCas9  \n",
      "GSM6745625 -0.249838                           1              dCas9  \n",
      "GSM6745626 -0.167241                           1           BE4alone  \n",
      "GSM6745627 -0.453903                           1           BE4alone  \n",
      "GSM6745628 -0.274132                           1           BE4alone  \n",
      "GSM6745629 -0.332436                           1          ABE8alone  \n",
      "GSM6745630 -0.385882                           1          ABE8alone  \n",
      "GSM6745631 -0.206110                           1          ABE8alone  \n",
      "GSM6745632 -0.196393                           0                 UT  \n",
      "GSM6745633 -0.191534                           0                 UT  \n",
      "GSM6745634 -0.342153                           0                 UT  \n",
      "GSM6745635 -0.361588                           0                 UT  \n",
      "GSM6745636 -0.361588                           0                 UT  \n",
      "GSM6745637 -0.337295                           0                 UT  \n",
      "\n",
      "[38 rows x 4801 columns]\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.decomposition import PCA\n",
    "import numpy as np\n",
    "\n",
    "df = pd.read_csv('../../data/GSE218462_raw_counts_GRCh38.p13_NCBI.tsv', sep='\\t')\n",
    "df = df.T\n",
    "df.columns = df.iloc[0]\n",
    "df = df[1:]\n",
    "\n",
    "scaler = StandardScaler()\n",
    "scaled_data = scaler.fit_transform(df)\n",
    "scaled_df = pd.DataFrame(scaled_data)\n",
    "\n",
    "scaled_df.columns = df.columns\n",
    "scaled_df.index = df.index\n",
    "unedited = ['GSM6745632', 'GSM6745633', 'GSM6745634', 'GSM6745635', 'GSM6745636', 'GSM6745637']\n",
    "scaled_df['Edited (1) or Unedited (0)'] = scaled_df.index.map(lambda gene: 0 if gene in unedited else 1)\n",
    "mechanisms = {\n",
    "    \"BE4\": [\"GSM6745599\", \"GSM6745600\", \"GSM6745601\", \"GSM6745611\", \"GSM6745612\", \"GSM6745613\"],\n",
    "    \"ABE8\": [\"GSM6745602\", \"GSM6745603\", \"GSM6745604\", \"GSM6745614\", \"GSM6745615\", \"GSM6745616\"],\n",
    "    \"Cas9\": [\"GSM6745605\", \"GSM6745606\", \"GSM6745607\", \"GSM6745617\", \"GSM6745618\", \"GSM6745619\"],\n",
    "    \"Utelectro\": [\"GSM6745608\", \"GSM6745609\", \"GSM6745610\", \"GSM6745620\", \"GSM6745621\", \"GSM6745622\"],\n",
    "    \"dCas9\": [\"GSM6745623\", \"GSM6745624\", \"GSM6745625\"],\n",
    "    \"BE4alone\": [\"GSM6745626\", \"GSM6745627\", \"GSM6745628\"],\n",
    "    \"ABE8alone\": [\"GSM6745629\", \"GSM6745630\", \"GSM6745631\"],\n",
    "    \"UT\": [\"GSM6745632\", \"GSM6745633\", \"GSM6745634\", \"GSM6745635\", \"GSM6745636\", \"GSM6745637\"]\n",
    "}\n",
    "\n",
    "# Inverting the dictionary to map gene code to its corresponding key\n",
    "mechanism_map = {gene: mechanism for mechanism, genes in mechanisms.items() for gene in genes}\n",
    "\n",
    "scaled_df['editing mechanism'] = scaled_df.index.map(mechanism_map)\n",
    "\n",
    "# print(scaled_df)\n",
    "\n",
    "txt_file_path = '../EDA_sj/relevant_genes_1.6_250.txt'\n",
    "with open(txt_file_path, 'r') as file:\n",
    "    txt_data = file.read()\n",
    "\n",
    "soham_gene_ids = txt_data.split(',')\n",
    "\n",
    "column_names = scaled_df.columns.tolist()\n",
    "\n",
    "column_names = [str(col) for col in scaled_df.columns.tolist()]\n",
    "filtered_column_names = [col for col in scaled_df.columns if str(col) in soham_gene_ids]\n",
    "scaled_df = scaled_df[filtered_column_names + ['Edited (1) or Unedited (0)', 'editing mechanism']]\n",
    "\n",
    "# Output the filtered DataFrame\n",
    "print(scaled_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/n1/ddw6p35j0310zh6cql00y29r0000gn/T/ipykernel_67196/4125254461.py:2: DtypeWarning: Columns (8,9) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  tsv_df = pd.read_csv(tsv_file_path, sep='\\t')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>Description</th>\n",
       "      <th>Synonyms</th>\n",
       "      <th>GeneType</th>\n",
       "      <th>EnsemblGeneID</th>\n",
       "      <th>Status</th>\n",
       "      <th>ChrAcc</th>\n",
       "      <th>ChrStart</th>\n",
       "      <th>ChrStop</th>\n",
       "      <th>Orientation</th>\n",
       "      <th>Length</th>\n",
       "      <th>GOFunctionID</th>\n",
       "      <th>GOProcessID</th>\n",
       "      <th>GOComponentID</th>\n",
       "      <th>GOFunction</th>\n",
       "      <th>GOProcess</th>\n",
       "      <th>GOComponent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>100287102</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>DEAD/H-box helicase 11 like 1 (pseudogene)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pseudo</td>\n",
       "      <td>ENSG00000290825</td>\n",
       "      <td>active</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>positive</td>\n",
       "      <td>1652</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>653635</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>WASP family homolog 7, pseudogene</td>\n",
       "      <td>FAM39F|WASH5P</td>\n",
       "      <td>pseudo</td>\n",
       "      <td>NaN</td>\n",
       "      <td>active</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>14362</td>\n",
       "      <td>29370</td>\n",
       "      <td>negative</td>\n",
       "      <td>1769</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>102466751</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>microRNA 6859-1</td>\n",
       "      <td>hsa-mir-6859-1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>ENSG00000278267</td>\n",
       "      <td>active</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>negative</td>\n",
       "      <td>68</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>107985730</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>MIR1302-2 host gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>active</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>29926</td>\n",
       "      <td>31295</td>\n",
       "      <td>positive</td>\n",
       "      <td>538</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>100302278</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>microRNA 1302-2</td>\n",
       "      <td>MIRN1302-2|hsa-mir-1302-2</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>ENSG00000284332</td>\n",
       "      <td>active</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "      <td>positive</td>\n",
       "      <td>138</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO:0035195</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>miRNA-mediated gene silencing</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39371</th>\n",
       "      <td>4541</td>\n",
       "      <td>ND6</td>\n",
       "      <td>NADH dehydrogenase subunit 6</td>\n",
       "      <td>MTND6</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>active</td>\n",
       "      <td>NC_012920.1</td>\n",
       "      <td>14149</td>\n",
       "      <td>14673</td>\n",
       "      <td>negative</td>\n",
       "      <td>525</td>\n",
       "      <td>GO:0008137</td>\n",
       "      <td>GO:0006120///GO:0009060///GO:0032981///GO:0035...</td>\n",
       "      <td>GO:0005739///GO:0005743///GO:0005747</td>\n",
       "      <td>NADH dehydrogenase (ubiquinone) activity</td>\n",
       "      <td>mitochondrial electron transport, NADH to ubiq...</td>\n",
       "      <td>mitochondrion///mitochondrial inner membrane//...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39372</th>\n",
       "      <td>4556</td>\n",
       "      <td>TRNE</td>\n",
       "      <td>tRNA-Glu</td>\n",
       "      <td>MTTE</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>active</td>\n",
       "      <td>NC_012920.1</td>\n",
       "      <td>14674</td>\n",
       "      <td>14742</td>\n",
       "      <td>negative</td>\n",
       "      <td>69</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39373</th>\n",
       "      <td>4519</td>\n",
       "      <td>CYTB</td>\n",
       "      <td>cytochrome b</td>\n",
       "      <td>MTCYB</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>active</td>\n",
       "      <td>NC_012920.1</td>\n",
       "      <td>14747</td>\n",
       "      <td>15887</td>\n",
       "      <td>positive</td>\n",
       "      <td>1141</td>\n",
       "      <td>GO:0008121///GO:0046872</td>\n",
       "      <td>GO:0006122///GO:0045333///GO:1902600</td>\n",
       "      <td>GO:0005739///GO:0005743///GO:0005750///GO:0016020</td>\n",
       "      <td>ubiquinol-cytochrome-c reductase activity///me...</td>\n",
       "      <td>mitochondrial electron transport, ubiquinol to...</td>\n",
       "      <td>mitochondrion///mitochondrial inner membrane//...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39374</th>\n",
       "      <td>4576</td>\n",
       "      <td>TRNT</td>\n",
       "      <td>tRNA-Thr</td>\n",
       "      <td>MTTT</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>active</td>\n",
       "      <td>NC_012920.1</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>positive</td>\n",
       "      <td>66</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39375</th>\n",
       "      <td>4571</td>\n",
       "      <td>TRNP</td>\n",
       "      <td>tRNA-Pro</td>\n",
       "      <td>MTTP</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>active</td>\n",
       "      <td>NC_012920.1</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>negative</td>\n",
       "      <td>68</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>39376 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          GeneID       Symbol                                 Description  \\\n",
       "0      100287102      DDX11L1  DEAD/H-box helicase 11 like 1 (pseudogene)   \n",
       "1         653635       WASH7P           WASP family homolog 7, pseudogene   \n",
       "2      102466751    MIR6859-1                             microRNA 6859-1   \n",
       "3      107985730  MIR1302-2HG                         MIR1302-2 host gene   \n",
       "4      100302278    MIR1302-2                             microRNA 1302-2   \n",
       "...          ...          ...                                         ...   \n",
       "39371       4541          ND6                NADH dehydrogenase subunit 6   \n",
       "39372       4556         TRNE                                    tRNA-Glu   \n",
       "39373       4519         CYTB                                cytochrome b   \n",
       "39374       4576         TRNT                                    tRNA-Thr   \n",
       "39375       4571         TRNP                                    tRNA-Pro   \n",
       "\n",
       "                        Synonyms        GeneType    EnsemblGeneID  Status  \\\n",
       "0                            NaN          pseudo  ENSG00000290825  active   \n",
       "1                  FAM39F|WASH5P          pseudo              NaN  active   \n",
       "2                 hsa-mir-6859-1           ncRNA  ENSG00000278267  active   \n",
       "3                            NaN           ncRNA              NaN  active   \n",
       "4      MIRN1302-2|hsa-mir-1302-2           ncRNA  ENSG00000284332  active   \n",
       "...                          ...             ...              ...     ...   \n",
       "39371                      MTND6  protein-coding              NaN  active   \n",
       "39372                       MTTE            tRNA              NaN  active   \n",
       "39373                      MTCYB  protein-coding              NaN  active   \n",
       "39374                       MTTT            tRNA              NaN  active   \n",
       "39375                       MTTP            tRNA              NaN  active   \n",
       "\n",
       "             ChrAcc ChrStart ChrStop Orientation  Length  \\\n",
       "0      NC_000001.11    11874   14409    positive    1652   \n",
       "1      NC_000001.11    14362   29370    negative    1769   \n",
       "2      NC_000001.11    17369   17436    negative      68   \n",
       "3      NC_000001.11    29926   31295    positive     538   \n",
       "4      NC_000001.11    30366   30503    positive     138   \n",
       "...             ...      ...     ...         ...     ...   \n",
       "39371   NC_012920.1    14149   14673    negative     525   \n",
       "39372   NC_012920.1    14674   14742    negative      69   \n",
       "39373   NC_012920.1    14747   15887    positive    1141   \n",
       "39374   NC_012920.1    15888   15953    positive      66   \n",
       "39375   NC_012920.1    15956   16023    negative      68   \n",
       "\n",
       "                  GOFunctionID  \\\n",
       "0                          NaN   \n",
       "1                          NaN   \n",
       "2                          NaN   \n",
       "3                          NaN   \n",
       "4                          NaN   \n",
       "...                        ...   \n",
       "39371               GO:0008137   \n",
       "39372                      NaN   \n",
       "39373  GO:0008121///GO:0046872   \n",
       "39374                      NaN   \n",
       "39375                      NaN   \n",
       "\n",
       "                                             GOProcessID  \\\n",
       "0                                                    NaN   \n",
       "1                                                    NaN   \n",
       "2                                                    NaN   \n",
       "3                                                    NaN   \n",
       "4                                             GO:0035195   \n",
       "...                                                  ...   \n",
       "39371  GO:0006120///GO:0009060///GO:0032981///GO:0035...   \n",
       "39372                                                NaN   \n",
       "39373               GO:0006122///GO:0045333///GO:1902600   \n",
       "39374                                                NaN   \n",
       "39375                                                NaN   \n",
       "\n",
       "                                           GOComponentID  \\\n",
       "0                                                    NaN   \n",
       "1                                                    NaN   \n",
       "2                                                    NaN   \n",
       "3                                                    NaN   \n",
       "4                                                    NaN   \n",
       "...                                                  ...   \n",
       "39371               GO:0005739///GO:0005743///GO:0005747   \n",
       "39372                                                NaN   \n",
       "39373  GO:0005739///GO:0005743///GO:0005750///GO:0016020   \n",
       "39374                                                NaN   \n",
       "39375                                                NaN   \n",
       "\n",
       "                                              GOFunction  \\\n",
       "0                                                    NaN   \n",
       "1                                                    NaN   \n",
       "2                                                    NaN   \n",
       "3                                                    NaN   \n",
       "4                                                    NaN   \n",
       "...                                                  ...   \n",
       "39371           NADH dehydrogenase (ubiquinone) activity   \n",
       "39372                                                NaN   \n",
       "39373  ubiquinol-cytochrome-c reductase activity///me...   \n",
       "39374                                                NaN   \n",
       "39375                                                NaN   \n",
       "\n",
       "                                               GOProcess  \\\n",
       "0                                                    NaN   \n",
       "1                                                    NaN   \n",
       "2                                                    NaN   \n",
       "3                                                    NaN   \n",
       "4                          miRNA-mediated gene silencing   \n",
       "...                                                  ...   \n",
       "39371  mitochondrial electron transport, NADH to ubiq...   \n",
       "39372                                                NaN   \n",
       "39373  mitochondrial electron transport, ubiquinol to...   \n",
       "39374                                                NaN   \n",
       "39375                                                NaN   \n",
       "\n",
       "                                             GOComponent  \n",
       "0                                                    NaN  \n",
       "1                                                    NaN  \n",
       "2                                                    NaN  \n",
       "3                                                    NaN  \n",
       "4                                                    NaN  \n",
       "...                                                  ...  \n",
       "39371  mitochondrion///mitochondrial inner membrane//...  \n",
       "39372                                                NaN  \n",
       "39373  mitochondrion///mitochondrial inner membrane//...  \n",
       "39374                                                NaN  \n",
       "39375                                                NaN  \n",
       "\n",
       "[39376 rows x 18 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tsv_file_path = '../../data/Human.GRCh38.p13.annot.tsv'\n",
    "tsv_df = pd.read_csv(tsv_file_path, sep='\\t')\n",
    "\n",
    "tsv_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Merged Data (first few rows):\n",
      "Index(['GSM6745599', 'GSM6745600', 'GSM6745601', 'GSM6745602', 'GSM6745603',\n",
      "       'GSM6745604', 'GSM6745605', 'GSM6745606', 'GSM6745607', 'GSM6745609',\n",
      "       'GSM6745610', 'GSM6745611', 'GSM6745612', 'GSM6745613', 'GSM6745614',\n",
      "       'GSM6745615', 'GSM6745616', 'GSM6745617', 'GSM6745618', 'GSM6745619',\n",
      "       'GSM6745620', 'GSM6745621', 'GSM6745622', 'GSM6745623', 'GSM6745624',\n",
      "       'GSM6745625', 'GSM6745626', 'GSM6745627', 'GSM6745628', 'GSM6745629',\n",
      "       'GSM6745630', 'GSM6745631', 'GSM6745632', 'GSM6745633', 'GSM6745634',\n",
      "       'GSM6745635', 'GSM6745636', 'GSM6745637', 'Symbol', 'Description',\n",
      "       'Synonyms', 'GeneType', 'EnsemblGeneID', 'Status', 'ChrAcc', 'ChrStart',\n",
      "       'ChrStop', 'Orientation', 'Length', 'GOFunctionID', 'GOProcessID',\n",
      "       'GOComponentID', 'GOFunction', 'GOProcess', 'GOComponent'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/n1/ddw6p35j0310zh6cql00y29r0000gn/T/ipykernel_67196/3509068322.py:2: DtypeWarning: Columns (8,9) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  metadata = pd.read_csv(metadata_file_path, sep='\\t')\n"
     ]
    }
   ],
   "source": [
    "metadata_file_path = \"../../data/Human.GRCh38.p13.annot.tsv\"\n",
    "metadata = pd.read_csv(metadata_file_path, sep='\\t')\n",
    "\n",
    "transposed_data = scaled_df.T\n",
    "transposed_data = transposed_data.reset_index()\n",
    "transposed_data.columns.values[0] = 'GeneID'\n",
    "\n",
    "\n",
    "merged_data = transposed_data.merge(metadata, on='GeneID', how='left').set_index('GeneID')\n",
    "merged_data = merged_data.iloc[:39378]\n",
    "\n",
    "print(\"Merged Data (first few rows):\")\n",
    "\n",
    "# Optionally, save to a new file\n",
    "output_file_path = \"merged_gene_expression_with_metadata.csv\"\n",
    "merged_data.to_csv(output_file_path)\n",
    "print(merged_data.columns)\n",
    "# print(f\"Merged data saved to {output_file_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['GSM6745599', 'GSM6745600', 'GSM6745601', 'GSM6745602', 'GSM6745603',\n",
      "       'GSM6745604', 'GSM6745605', 'GSM6745606', 'GSM6745607', 'GSM6745609',\n",
      "       'GSM6745610', 'GSM6745611', 'GSM6745612', 'GSM6745613', 'GSM6745614',\n",
      "       'GSM6745615', 'GSM6745616', 'GSM6745617', 'GSM6745618', 'GSM6745619',\n",
      "       'GSM6745620', 'GSM6745621', 'GSM6745622', 'GSM6745623', 'GSM6745624',\n",
      "       'GSM6745625', 'GSM6745626', 'GSM6745627', 'GSM6745628', 'GSM6745629',\n",
      "       'GSM6745630', 'GSM6745631', 'GSM6745632', 'GSM6745633', 'GSM6745634',\n",
      "       'GSM6745635', 'GSM6745636', 'GSM6745637', 'Symbol', 'Description',\n",
      "       'Synonyms', 'GeneType', 'EnsemblGeneID', 'Status', 'ChrAcc', 'ChrStart',\n",
      "       'ChrStop', 'Orientation', 'Length', 'GOFunctionID', 'GOProcessID',\n",
      "       'GOComponentID', 'GOFunction', 'GOProcess', 'GOComponent'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "# merged_data = merged_data.iloc[:, :-6]\n",
    "# merged_data = merged_data.drop(columns=['EnsemblGeneID'])\n",
    "merged_data\n",
    "output_file_path = \"testing.csv\"\n",
    "merged_data.to_csv(output_file_path)\n",
    "print(merged_data.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.int64(1221)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data_nulls = merged_data['Description'].isnull().sum() + merged_data['Description'].str.contains('uncharacterized', case=False, na=False).sum()\n",
    "merged_data_nulls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows: 4656, Number of columns: 55\n",
      "         GeneID GSM6745599 GSM6745600 GSM6745601 GSM6745602 GSM6745603  \\\n",
      "0     113219467  -0.158632   1.379276   5.355088   2.002351   0.679553   \n",
      "1         57801   1.070933   0.767235   3.457126   0.897391   1.070933   \n",
      "2          9636    1.15215   0.761592   2.062135   0.336385   0.673147   \n",
      "3     100288175   0.499194  -0.765431   1.706335  -0.075635    0.61416   \n",
      "4     102465434  -0.342997  -0.342997  -0.342997  -0.342997  -0.342997   \n",
      "...         ...        ...        ...        ...        ...        ...   \n",
      "4651  105377243  -0.222988  -0.222988  -0.222988  -0.222988  -0.222988   \n",
      "4652  107987359  -0.164399  -0.164399  -0.164399  -0.164399  -0.164399   \n",
      "4653       4572  -0.135784   0.567824   3.460437    1.42779   2.131398   \n",
      "4654       4512  -0.237304   1.302546   5.055768   2.327118   0.856884   \n",
      "4655       4574  -0.162382   1.353526    5.39109   1.941426   0.658735   \n",
      "\n",
      "     GSM6745604 GSM6745605 GSM6745606 GSM6745607  ...  ChrStart   ChrStop  \\\n",
      "0     -0.213027     0.0095  -0.393521   -0.27484  ...    632615    632685   \n",
      "1      1.548171   0.810621  -0.274013   0.289997  ...    998962   1001052   \n",
      "2      0.828458  -0.304008   -0.72496  -0.605817  ...   1013497   1014540   \n",
      "3     -0.018153    1.01654  -0.650465   0.384228  ...   1059734   1066453   \n",
      "4     -0.342997  -0.342997  -0.342997  -0.342997  ...   1296110   1296170   \n",
      "...         ...        ...        ...        ...  ...       ...       ...   \n",
      "4651  -0.222988  -0.222988  -0.222988  -0.222988  ...  26196209  26235374   \n",
      "4652  -0.164399  -0.164399  -0.164399  -0.164399  ...     35630     43681   \n",
      "4653   1.975041   1.349611   0.098752   0.724182  ...      4329      4400   \n",
      "4654  -0.052044  -0.139682  -0.558824  -0.258147  ...      5904      7445   \n",
      "4655   -0.20611  -0.016622  -0.381023  -0.269273  ...      7446      7514   \n",
      "\n",
      "     Orientation  Length                                       GOFunctionID  \\\n",
      "0       negative    71.0                                         GO:0030533   \n",
      "1       negative  1920.0  GO:0000978///GO:0000981///GO:0005515///GO:0046983   \n",
      "2       positive   637.0  GO:0003735///GO:0005178///GO:0005515///GO:0031...   \n",
      "3       positive   974.0                                                NaN   \n",
      "4       negative    61.0                                                NaN   \n",
      "...          ...     ...                                                ...   \n",
      "4651    negative  2725.0                                                NaN   \n",
      "4652    negative   745.0                                                NaN   \n",
      "4653    negative    72.0                                                NaN   \n",
      "4654    positive  1542.0  GO:0004129///GO:0005515///GO:0020037///GO:0046872   \n",
      "4655    negative    69.0                                                NaN   \n",
      "\n",
      "                                            GOProcessID  \\\n",
      "0                                            GO:0006412   \n",
      "1     GO:0006357///GO:0007399///GO:0009952///GO:0030154   \n",
      "2     GO:0007229///GO:0009615///GO:0016567///GO:0019...   \n",
      "3                                                   NaN   \n",
      "4                                                   NaN   \n",
      "...                                                 ...   \n",
      "4651                                                NaN   \n",
      "4652                                                NaN   \n",
      "4653                                                NaN   \n",
      "4654  GO:0006123///GO:0006979///GO:0007568///GO:0009...   \n",
      "4655                                                NaN   \n",
      "\n",
      "                                          GOComponentID  \\\n",
      "0                                                   NaN   \n",
      "1                               GO:0000785///GO:0005634   \n",
      "2     GO:0005576///GO:0005634///GO:0005654///GO:0005...   \n",
      "3                                                   NaN   \n",
      "4                                                   NaN   \n",
      "...                                                 ...   \n",
      "4651                                                NaN   \n",
      "4652                                                NaN   \n",
      "4653                                                NaN   \n",
      "4654  GO:0005743///GO:0005750///GO:0005751///GO:0045277   \n",
      "4655                                                NaN   \n",
      "\n",
      "                                             GOFunction  \\\n",
      "0             triplet codon-amino acid adaptor activity   \n",
      "1     RNA polymerase II cis-regulatory region sequen...   \n",
      "2     structural constituent of ribosome///integrin ...   \n",
      "3                                                   NaN   \n",
      "4                                                   NaN   \n",
      "...                                                 ...   \n",
      "4651                                                NaN   \n",
      "4652                                                NaN   \n",
      "4653                                                NaN   \n",
      "4654  cytochrome-c oxidase activity///protein bindin...   \n",
      "4655                                                NaN   \n",
      "\n",
      "                                              GOProcess  \\\n",
      "0                                           translation   \n",
      "1     regulation of transcription by RNA polymerase ...   \n",
      "2     integrin-mediated signaling pathway///response...   \n",
      "3                                                   NaN   \n",
      "4                                                   NaN   \n",
      "...                                                 ...   \n",
      "4651                                                NaN   \n",
      "4652                                                NaN   \n",
      "4653                                                NaN   \n",
      "4654  mitochondrial electron transport, cytochrome c...   \n",
      "4655                                                NaN   \n",
      "\n",
      "                                            GOComponent  \n",
      "0                                                   NaN  \n",
      "1                                   chromatin///nucleus  \n",
      "2     extracellular region///nucleus///nucleoplasm//...  \n",
      "3                                                   NaN  \n",
      "4                                                   NaN  \n",
      "...                                                 ...  \n",
      "4651                                                NaN  \n",
      "4652                                                NaN  \n",
      "4653                                                NaN  \n",
      "4654  mitochondrial inner membrane///mitochondrial r...  \n",
      "4655                                                NaN  \n",
      "\n",
      "[4656 rows x 56 columns]\n"
     ]
    }
   ],
   "source": [
    "# Delete any genes with missing descriptions\n",
    "merged_data = merged_data.dropna(subset=['Description'])\n",
    "print(f\"Number of rows: {merged_data.shape[0]}, Number of columns: {merged_data.shape[1]}\")\n",
    "merged_data.reset_index(inplace=True) # Apparently, I had GeneID as the index, so I reset it to a column cuz it's easier to work with\n",
    "print(merged_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Relevant genes with contributions for PC1:\n",
      "            Loading  Contribution (%)\n",
      "GeneID                               \n",
      "102724659  0.520019         12.013637\n",
      "105376737  0.404821          9.352300\n",
      "6060       0.336847          7.781946\n",
      "26869      0.281715          6.508251\n",
      "9636       0.264705          6.115286\n",
      "55092      0.248511          5.741164\n",
      "644068     0.187465          4.330861\n",
      "26279      0.159415          3.682860\n",
      "29943      0.151901          3.509253\n",
      "7799       0.142350          3.288610\n",
      "57801      0.138742          3.205249\n",
      "54751      0.132867          3.069541\n",
      "107984872  0.124788          2.882886\n",
      "127707     0.118198          2.730653\n",
      "155184     0.098019          2.264469\n",
      "26870      0.089438          2.066230\n",
      "105376691  0.082548          1.907053\n",
      "8510       0.077518          1.790838\n",
      "27129      0.076625          1.770211\n",
      "112268220  0.074872          1.729708\n",
      "63036      0.067704          1.564120\n",
      "105376809  0.061128          1.412197\n",
      "\n",
      "Relevant genes with contributions for PC2:\n",
      "            Loading  Contribution (%)\n",
      "GeneID                               \n",
      "127707     0.519959         11.637367\n",
      "26279      0.477684         10.691192\n",
      "644068     0.345249          7.727135\n",
      "51154      0.264382          5.917209\n",
      "5320       0.186603          4.176410\n",
      "102465434  0.180050          4.029758\n",
      "100288175  0.174027          3.894943\n",
      "284661     0.142808          3.196227\n",
      "6060       0.138239          3.093966\n",
      "30814      0.126037          2.820871\n",
      "109623456  0.124950          2.796548\n",
      "55092      0.115196          2.578234\n",
      "102724659  0.108266          2.423139\n",
      "57801      0.107846          2.413729\n",
      "105376737  0.101363          2.268640\n",
      "107984872  0.100529          2.249982\n",
      "8510       0.098410          2.202538\n",
      "26870      0.091398          2.045605\n",
      "155184     0.089009          1.992133\n",
      "29943      0.087516          1.958731\n",
      "106614088  0.086446          1.934785\n",
      "9636       0.080090          1.792515\n",
      "1187       0.071271          1.595146\n",
      "26869      0.069904          1.564547\n",
      "1969       0.067589          1.512723\n",
      "\n",
      "Relevant genes with contributions for PC3:\n",
      "            Loading  Contribution (%)\n",
      "GeneID                               \n",
      "26279      0.816549         32.481680\n",
      "127707     0.430822         17.137748\n",
      "644068     0.332662         13.233043\n",
      "51154      0.094422          3.756033\n",
      "9636       0.086821          3.453663\n",
      "109623456  0.047165          1.876189\n",
      "8510       0.046453          1.847878\n",
      "102465434  0.041936          1.668164\n",
      "113219467  0.037886          1.507085\n",
      "102724659  0.035129          1.397394\n",
      "112268220  0.031784          1.264344\n",
      "284661     0.031272          1.243966\n",
      "100288175  0.031087          1.236596\n",
      "105376691  0.030426          1.210333\n",
      "10630      0.025615          1.018942\n",
      "26829      0.025398          1.010320\n",
      "105376809  0.024182          0.961946\n",
      "26870      0.023542          0.936474\n",
      "106614088  0.022258          0.885388\n",
      "7799       0.021754          0.865361\n",
      "105376739  0.020767          0.826106\n",
      "\n",
      "Relevant genes with contributions for PC4:\n",
      "            Loading  Contribution (%)\n",
      "GeneID                               \n",
      "644068     0.748155         32.141369\n",
      "127707     0.631450         27.127647\n",
      "102724659  0.078119          3.356038\n",
      "9636       0.073168          3.143349\n",
      "102465434  0.073154          3.142757\n",
      "109623456  0.067268          2.889874\n",
      "8510       0.057446          2.467935\n",
      "100288175  0.053173          2.284376\n",
      "105376737  0.050237          2.158211\n",
      "26869      0.039192          1.683725\n",
      "105376809  0.031859          1.368685\n",
      "113219467  0.030004          1.289016\n",
      "284661     0.029939          1.286203\n",
      "7799       0.029533          1.268762\n",
      "107984872  0.028812          1.237768\n",
      "57801      0.028171          1.210268\n",
      "27129      0.026681          1.146254\n",
      "\n",
      "Relevant genes with contributions for PC5:\n",
      "            Loading  Contribution (%)\n",
      "GeneID                               \n",
      "9636       0.511579          9.987875\n",
      "8510       0.338439          6.607548\n",
      "644068     0.257498          5.027290\n",
      "109623456  0.231957          4.528647\n",
      "102724659  0.224368          4.380467\n",
      "102465434  0.214545          4.188701\n",
      "1187       0.192750          3.763168\n",
      "105376809  0.182925          3.571362\n",
      "26870      0.175031          3.417245\n",
      "100288175  0.173476          3.386870\n",
      "5320       0.153312          2.993198\n",
      "113219467  0.138796          2.709805\n",
      "7799       0.136807          2.670974\n",
      "1969       0.135587          2.647155\n",
      "26829      0.130767          2.553036\n",
      "26869      0.128051          2.500010\n",
      "106614088  0.126990          2.479307\n",
      "26871      0.123217          2.405635\n",
      "112268220  0.116530          2.275087\n",
      "105376805  0.109840          2.144481\n",
      "105376691  0.101661          1.984781\n",
      "105376739  0.098179          1.916808\n",
      "30814      0.096679          1.887521\n",
      "63036      0.094613          1.847187\n",
      "10630      0.094459          1.844175\n",
      "27129      0.093716          1.829677\n",
      "29943      0.091090          1.778415\n",
      "155184     0.088721          1.732144\n",
      "\n",
      "Relevant genes with contributions for PC6:\n",
      "            Loading  Contribution (%)\n",
      "GeneID                               \n",
      "51154      0.742577         22.183944\n",
      "5320       0.330176          9.863762\n",
      "127707     0.291262          8.701256\n",
      "9636       0.274480          8.199882\n",
      "644068     0.240527          7.185580\n",
      "26279      0.163028          4.870351\n",
      "30814      0.147184          4.397012\n",
      "8510       0.135277          4.041297\n",
      "109623456  0.098638          2.946754\n",
      "102465434  0.097516          2.913207\n",
      "100288175  0.090182          2.694121\n",
      "1187       0.061770          1.845335\n",
      "284661     0.060536          1.808476\n",
      "102724659  0.058208          1.738937\n",
      "105376737  0.053099          1.586296\n",
      "26829      0.042231          1.261622\n",
      "1969       0.037067          1.107348\n",
      "106614088  0.035278          1.053909\n",
      "57801      0.034501          1.030707\n",
      "\n",
      "Relevant genes with contributions for PC7:\n",
      "            Loading  Contribution (%)\n",
      "GeneID                               \n",
      "8510       0.593315         18.650239\n",
      "9636       0.587611         18.470923\n",
      "51154      0.323239         10.160685\n",
      "109623456  0.273905          8.609914\n",
      "100288175  0.200234          6.294133\n",
      "105376737  0.175921          5.529909\n",
      "284661     0.105229          3.307752\n",
      "127707     0.086626          2.722998\n",
      "644068     0.075920          2.386473\n",
      "102465434  0.061727          1.940330\n",
      "1187       0.051749          1.626675\n",
      "26870      0.048415          1.521872\n",
      "112268220  0.045940          1.444076\n",
      "105376809  0.045461          1.429007\n",
      "26279      0.043333          1.362137\n",
      "26829      0.041524          1.305276\n",
      "106614088  0.039101          1.229089\n",
      "30814      0.036953          1.161577\n",
      "\n",
      "Relevant genes with contributions for PC8:\n",
      "            Loading  Contribution (%)\n",
      "GeneID                               \n",
      "102465434  0.910703         38.184730\n",
      "8510       0.279963         11.738518\n",
      "109623456  0.152361          6.388330\n",
      "9636       0.140583          5.894474\n",
      "100288175  0.136041          5.704055\n",
      "51154      0.077311          3.241552\n",
      "102724659  0.070281          2.946793\n",
      "29943      0.069340          2.907355\n",
      "644068     0.059761          2.505702\n",
      "127707     0.044934          1.884035\n",
      "26870      0.038814          1.627447\n",
      "26279      0.034116          1.430463\n",
      "55092      0.032359          1.356769\n",
      "6060       0.025317          1.061509\n",
      "54751      0.023917          1.002794\n",
      "155184     0.023214          0.973340\n",
      "105376809  0.022986          0.963770\n",
      "\n",
      "Relevant genes with contributions for PC9:\n",
      "            Loading  Contribution (%)\n",
      "GeneID                               \n",
      "109623456  0.798222         29.387728\n",
      "8510       0.466814         17.186427\n",
      "100288175  0.236697          8.714357\n",
      "9636       0.184723          6.800840\n",
      "102724659  0.104755          3.856708\n",
      "29943      0.101630          3.741645\n",
      "284661     0.092543          3.407123\n",
      "105376737  0.071655          2.638092\n",
      "107984872  0.046131          1.698397\n",
      "105376805  0.045772          1.685159\n",
      "644068     0.043507          1.601768\n",
      "6060       0.038216          1.406976\n",
      "155184     0.036925          1.359453\n",
      "26871      0.035814          1.318541\n",
      "102465434  0.033641          1.238543\n",
      "51154      0.033340          1.227471\n",
      "112268220  0.031932          1.175628\n",
      "55092      0.029510          1.086464\n",
      "\n",
      "Relevant genes with contributions for PC10:\n",
      "            Loading  Contribution (%)\n",
      "GeneID                               \n",
      "100288175  0.850200         31.736548\n",
      "109623456  0.305962         11.421056\n",
      "9636       0.305349         11.398178\n",
      "102724659  0.142003          5.300741\n",
      "284661     0.140060          5.228221\n",
      "8510       0.136269          5.086696\n",
      "29943      0.101058          3.772330\n",
      "26870      0.059484          2.220431\n",
      "6060       0.044337          1.655037\n",
      "26869      0.042896          1.601224\n",
      "102465434  0.039888          1.488966\n",
      "155184     0.037030          1.382258\n",
      "26871      0.036844          1.375313\n",
      "106614088  0.035958          1.342254\n",
      "105376805  0.034412          1.284554\n",
      "55092      0.033191          1.238978\n",
      "26279      0.030656          1.144330\n",
      "112268220  0.030035          1.121155\n",
      "\n",
      "Relevant genes with contributions for PC11:\n",
      "            Loading  Contribution (%)\n",
      "GeneID                               \n",
      "5320       0.783644         26.085389\n",
      "51154      0.415669         13.836492\n",
      "30814      0.237014          7.889550\n",
      "284661     0.222573          7.408850\n",
      "8510       0.179957          5.990294\n",
      "109623456  0.132127          4.398166\n",
      "102465434  0.107466          3.577250\n",
      "26279      0.107130          3.566051\n",
      "644068     0.068411          2.277201\n",
      "9636       0.062331          2.074817\n",
      "29943      0.062015          2.064303\n",
      "127707     0.057630          1.918343\n",
      "105376809  0.053968          1.796447\n",
      "7799       0.050474          1.680137\n",
      "10630      0.049039          1.632381\n",
      "105376691  0.047912          1.594867\n",
      "1187       0.034104          1.135234\n",
      "\n",
      "Relevant genes with contributions for PC12:\n",
      "            Loading  Contribution (%)\n",
      "GeneID                               \n",
      "284661     0.861634         26.340957\n",
      "5320       0.209456          6.403259\n",
      "51154      0.202843          6.201109\n",
      "8510       0.193753          5.923216\n",
      "105376691  0.168406          5.148339\n",
      "9636       0.115243          3.523082\n",
      "26870      0.095235          2.911418\n",
      "105376737  0.091572          2.799454\n",
      "100288175  0.085451          2.612329\n",
      "27129      0.085229          2.605516\n",
      "30814      0.078428          2.397608\n",
      "1187       0.077464          2.368132\n",
      "10630      0.075134          2.296904\n",
      "102465434  0.066289          2.026525\n",
      "26829      0.065345          1.997660\n",
      "105376809  0.064214          1.963073\n",
      "7799       0.062420          1.908243\n",
      "109623456  0.061578          1.882492\n",
      "54751      0.057046          1.743939\n",
      "63036      0.055195          1.687357\n",
      "105376739  0.050220          1.535285\n",
      "55092      0.048854          1.493517\n",
      "113219467  0.040540          1.239354\n",
      "\n",
      "Relevant genes with contributions for PC13:\n",
      "            Loading  Contribution (%)\n",
      "GeneID                               \n",
      "105376737  0.586915         16.532616\n",
      "102724659  0.468495         13.196886\n",
      "29943      0.377713         10.639664\n",
      "26869      0.308203          8.681668\n",
      "8510       0.171030          4.817695\n",
      "100288175  0.159725          4.499240\n",
      "1187       0.154114          4.341173\n",
      "284661     0.150104          4.228239\n",
      "26870      0.142449          4.012588\n",
      "112268220  0.127776          3.599288\n",
      "105376805  0.122954          3.463461\n",
      "105376691  0.105840          2.981376\n",
      "26829      0.087664          2.469385\n",
      "30814      0.078965          2.224331\n",
      "113219467  0.064100          1.805624\n",
      "5320       0.059765          1.683512\n",
      "\n",
      "Relevant genes with contributions for PC14:\n",
      "            Loading  Contribution (%)\n",
      "GeneID                               \n",
      "105376809  0.746528         30.503210\n",
      "26870      0.600824         24.549740\n",
      "105376737  0.204410          8.352217\n",
      "30814      0.092131          3.764484\n",
      "5320       0.065351          2.670249\n",
      "1187       0.063745          2.604634\n",
      "10630      0.060085          2.455090\n",
      "102724659  0.055088          2.250918\n",
      "26869      0.051863          2.119125\n",
      "26829      0.045937          1.876978\n",
      "7799       0.043821          1.790546\n",
      "26871      0.043809          1.790058\n",
      "105376691  0.038547          1.575030\n",
      "8510       0.031112          1.271238\n",
      "109623456  0.030826          1.259536\n",
      "27129      0.026414          1.079277\n",
      "\n",
      "Relevant genes with contributions for PC15:\n",
      "            Loading  Contribution (%)\n",
      "GeneID                               \n",
      "26870      0.558636         17.451259\n",
      "29943      0.536425         16.757390\n",
      "105376809  0.440902         13.773345\n",
      "102724659  0.355670         11.110785\n",
      "26869      0.124528          3.890123\n",
      "105376737  0.120736          3.771662\n",
      "105376691  0.094634          2.956269\n",
      "26829      0.077935          2.434620\n",
      "105376739  0.062438          1.950496\n",
      "54751      0.056240          1.756871\n",
      "5320       0.050312          1.571698\n",
      "30814      0.049112          1.534208\n",
      "105376805  0.046072          1.439255\n",
      "106614088  0.042676          1.333173\n",
      "107984872  0.040870          1.276748\n",
      "284661     0.040761          1.273320\n",
      "10630      0.039410          1.231120\n",
      "1187       0.037741          1.178988\n",
      "57801      0.035022          1.094067\n",
      "55092      0.034381          1.074034\n",
      "26871      0.034266          1.070440\n"
     ]
    }
   ],
   "source": [
    "expression_cols = merged_data.columns[1:39]\n",
    "expression_data = merged_data[expression_cols].apply(pd.to_numeric, errors='coerce')\n",
    "\n",
    "pca_target_variance = 0.75\n",
    "pca_full = PCA()\n",
    "expression_data_filled = expression_data.apply(lambda row: row.bfill().ffill(), axis=1)\n",
    "nan_counts = expression_data.isna().sum().sum()\n",
    "nan_counts\n",
    "\n",
    "pca_full.fit(expression_data)\n",
    "\n",
    "# Determine number of components needed to explain 75% variance\n",
    "cumulative_variance = pca_full.explained_variance_ratio_.cumsum()\n",
    "n_components_75 = np.argmax(cumulative_variance >= pca_target_variance) + 1\n",
    "\n",
    "# Apply PCA with the determined number of components\n",
    "pca_final = PCA(n_components=n_components_75)\n",
    "pca_result_final = pca_final.fit_transform(expression_data)\n",
    "\n",
    "# Retrieve top contributing genes for each principal component\n",
    "components_df = pd.DataFrame(\n",
    "    pca_final.components_,\n",
    "    columns=merged_data['GeneID'][:expression_data.shape[1]],\n",
    "    index=[f\"PC{i+1}\" for i in range(n_components_75)]\n",
    ")\n",
    "\n",
    "def get_relevant_genes_with_contribution(pc, cumulative_threshold=0.9):\n",
    "    # Absolute loadings for the specified principal component\n",
    "    loadings = components_df.iloc[pc-1].abs().sort_values(ascending=False)\n",
    "\n",
    "    # Calculate cumulative variance contribution by each gene\n",
    "    cumulative_variance = loadings.cumsum() / loadings.sum()\n",
    "\n",
    "    # Select genes up to the cumulative threshold\n",
    "    relevant_genes = loadings[cumulative_variance <= cumulative_threshold]\n",
    "\n",
    "    # Calculate the contribution percentage for each gene\n",
    "    contribution_percentages = (relevant_genes / loadings.sum()) * 100\n",
    "\n",
    "    return pd.DataFrame({\n",
    "        'Loading': relevant_genes,\n",
    "        'Contribution (%)': contribution_percentages\n",
    "    })\n",
    "\n",
    "# Example usage\n",
    "for pc in range(1, n_components_75 + 1):\n",
    "    print(f\"\\nRelevant genes with contributions for PC{pc}:\")\n",
    "    print(get_relevant_genes_with_contribution(pc))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/n1/ddw6p35j0310zh6cql00y29r0000gn/T/ipykernel_67196/281396425.py:8: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  data = data.applymap(lambda x: np.random.rand())\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'PC1 & PC2': 18,\n",
       " 'PC1 & PC3': 12,\n",
       " 'PC1 & PC4': 15,\n",
       " 'PC1 & PC5': 16,\n",
       " 'PC1 & PC6': 17,\n",
       " 'PC1 & PC7': 18,\n",
       " 'PC1 & PC8': 14,\n",
       " 'PC1 & PC9': 14,\n",
       " 'PC1 & PC10': 15,\n",
       " 'PC1 & PC11': 12,\n",
       " 'PC1 & PC12': 13,\n",
       " 'PC1 & PC13': 16,\n",
       " 'PC1 & PC14': 15,\n",
       " 'PC1 & PC15': 15,\n",
       " 'PC2 & PC3': 16,\n",
       " 'PC2 & PC4': 20,\n",
       " 'PC2 & PC5': 19,\n",
       " 'PC2 & PC6': 20,\n",
       " 'PC2 & PC7': 21,\n",
       " 'PC2 & PC8': 16,\n",
       " 'PC2 & PC9': 14,\n",
       " 'PC2 & PC10': 15,\n",
       " 'PC2 & PC11': 17,\n",
       " 'PC2 & PC12': 16,\n",
       " 'PC2 & PC13': 18,\n",
       " 'PC2 & PC14': 18,\n",
       " 'PC2 & PC15': 17,\n",
       " 'PC3 & PC4': 16,\n",
       " 'PC3 & PC5': 14,\n",
       " 'PC3 & PC6': 16,\n",
       " 'PC3 & PC7': 16,\n",
       " 'PC3 & PC8': 10,\n",
       " 'PC3 & PC9': 15,\n",
       " 'PC3 & PC10': 15,\n",
       " 'PC3 & PC11': 15,\n",
       " 'PC3 & PC12': 15,\n",
       " 'PC3 & PC13': 16,\n",
       " 'PC3 & PC14': 13,\n",
       " 'PC3 & PC15': 16,\n",
       " 'PC4 & PC5': 14,\n",
       " 'PC4 & PC6': 18,\n",
       " 'PC4 & PC7': 20,\n",
       " 'PC4 & PC8': 16,\n",
       " 'PC4 & PC9': 16,\n",
       " 'PC4 & PC10': 16,\n",
       " 'PC4 & PC11': 15,\n",
       " 'PC4 & PC12': 15,\n",
       " 'PC4 & PC13': 18,\n",
       " 'PC4 & PC14': 17,\n",
       " 'PC4 & PC15': 16,\n",
       " 'PC5 & PC6': 17,\n",
       " 'PC5 & PC7': 15,\n",
       " 'PC5 & PC8': 13,\n",
       " 'PC5 & PC9': 12,\n",
       " 'PC5 & PC10': 15,\n",
       " 'PC5 & PC11': 11,\n",
       " 'PC5 & PC12': 16,\n",
       " 'PC5 & PC13': 17,\n",
       " 'PC5 & PC14': 14,\n",
       " 'PC5 & PC15': 19,\n",
       " 'PC6 & PC7': 18,\n",
       " 'PC6 & PC8': 17,\n",
       " 'PC6 & PC9': 13,\n",
       " 'PC6 & PC10': 15,\n",
       " 'PC6 & PC11': 14,\n",
       " 'PC6 & PC12': 18,\n",
       " 'PC6 & PC13': 19,\n",
       " 'PC6 & PC14': 19,\n",
       " 'PC6 & PC15': 18,\n",
       " 'PC7 & PC8': 16,\n",
       " 'PC7 & PC9': 17,\n",
       " 'PC7 & PC10': 15,\n",
       " 'PC7 & PC11': 16,\n",
       " 'PC7 & PC12': 15,\n",
       " 'PC7 & PC13': 16,\n",
       " 'PC7 & PC14': 18,\n",
       " 'PC7 & PC15': 16,\n",
       " 'PC8 & PC9': 12,\n",
       " 'PC8 & PC10': 14,\n",
       " 'PC8 & PC11': 13,\n",
       " 'PC8 & PC12': 13,\n",
       " 'PC8 & PC13': 14,\n",
       " 'PC8 & PC14': 17,\n",
       " 'PC8 & PC15': 17,\n",
       " 'PC9 & PC10': 17,\n",
       " 'PC9 & PC11': 13,\n",
       " 'PC9 & PC12': 15,\n",
       " 'PC9 & PC13': 14,\n",
       " 'PC9 & PC14': 13,\n",
       " 'PC9 & PC15': 15,\n",
       " 'PC10 & PC11': 13,\n",
       " 'PC10 & PC12': 15,\n",
       " 'PC10 & PC13': 16,\n",
       " 'PC10 & PC14': 16,\n",
       " 'PC10 & PC15': 17,\n",
       " 'PC11 & PC12': 12,\n",
       " 'PC11 & PC13': 16,\n",
       " 'PC11 & PC14': 14,\n",
       " 'PC11 & PC15': 13,\n",
       " 'PC12 & PC13': 14,\n",
       " 'PC12 & PC14': 17,\n",
       " 'PC12 & PC15': 15,\n",
       " 'PC13 & PC14': 15,\n",
       " 'PC13 & PC15': 18,\n",
       " 'PC14 & PC15': 19}"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Assuming 'merged_data' is already loaded with 'GeneID' column and relevant expression data columns\n",
    "# Simulating a smaller set for demo purposes; in a real scenario, we would use the actual 'merged_data' DataFrame\n",
    "\n",
    "# Generate sample data similar to what was described\n",
    "gene_ids = [f'Gene_{i}' for i in range(1, 51)]\n",
    "pc_columns = [f'Exp_{i}' for i in range(1, 40)]\n",
    "data = pd.DataFrame(index=gene_ids, columns=pc_columns)\n",
    "data = data.applymap(lambda x: np.random.rand())\n",
    "\n",
    "# Simulating 'GeneID' in merged_data to create a column similar to the description\n",
    "data = data.reset_index().rename(columns={'index': 'GeneID'})\n",
    "\n",
    "# Selecting expression data columns\n",
    "expression_cols = data.columns[1:]\n",
    "expression_data = data[expression_cols].apply(pd.to_numeric, errors='coerce')\n",
    "\n",
    "# PCA process as described\n",
    "pca_target_variance = 0.75\n",
    "pca_full = PCA()\n",
    "expression_data_filled = expression_data.apply(lambda row: row.bfill().ffill(), axis=1)\n",
    "\n",
    "# Fitting PCA\n",
    "pca_full.fit(expression_data)\n",
    "\n",
    "# Determine number of components to explain 75% variance\n",
    "cumulative_variance = pca_full.explained_variance_ratio_.cumsum()\n",
    "n_components_75 = np.argmax(cumulative_variance >= pca_target_variance) + 1\n",
    "\n",
    "# Apply PCA with determined number of components\n",
    "pca_final = PCA(n_components=n_components_75)\n",
    "pca_result_final = pca_final.fit_transform(expression_data)\n",
    "\n",
    "# Retrieve top contributing genes for each principal component\n",
    "components_df = pd.DataFrame(\n",
    "    pca_final.components_,\n",
    "    columns=data['GeneID'][:expression_data.shape[1]],\n",
    "    index=[f\"PC{i+1}\" for i in range(n_components_75)]\n",
    ")\n",
    "\n",
    "def get_relevant_genes_with_contribution(pc, cumulative_threshold=0.9):\n",
    "    # Absolute loadings for the specified principal component\n",
    "    loadings = components_df.iloc[pc-1].abs().sort_values(ascending=False)\n",
    "    cumulative_variance = loadings.cumsum() / loadings.sum()\n",
    "    relevant_genes = loadings[cumulative_variance <= cumulative_threshold]\n",
    "    return relevant_genes.index.tolist()\n",
    "\n",
    "# Collect relevant genes for each PC\n",
    "relevant_genes_per_pc = {f\"PC{pc}\": get_relevant_genes_with_contribution(pc) for pc in range(1, n_components_75 + 1)}\n",
    "\n",
    "# Calculate overlaps between PC groups\n",
    "overlap_counts = {}\n",
    "for pc_i in range(1, n_components_75 + 1):\n",
    "    pc_i_genes = set(relevant_genes_per_pc[f\"PC{pc_i}\"])\n",
    "    for pc_j in range(pc_i + 1, n_components_75 + 1):\n",
    "        pc_j_genes = set(relevant_genes_per_pc[f\"PC{pc_j}\"])\n",
    "        overlap_counts[f\"PC{pc_i} & PC{pc_j}\"] = len(pc_i_genes & pc_j_genes)\n",
    "\n",
    "overlap_counts\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'plt' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[13], line 36\u001b[0m\n\u001b[1;32m     31\u001b[0m     overlap_percentage_df\u001b[38;5;241m.\u001b[39mloc[pc_j, pc_i] \u001b[38;5;241m=\u001b[39m (\n\u001b[1;32m     32\u001b[0m         percent  \u001b[38;5;66;03m# Mirror the percentage for both (i, j) and (j, i)\u001b[39;00m\n\u001b[1;32m     33\u001b[0m     )\n\u001b[1;32m     35\u001b[0m \u001b[38;5;66;03m# Set up the matplotlib figure\u001b[39;00m\n\u001b[0;32m---> 36\u001b[0m \u001b[43mplt\u001b[49m\u001b[38;5;241m.\u001b[39mfigure(figsize\u001b[38;5;241m=\u001b[39m(\u001b[38;5;241m10\u001b[39m, \u001b[38;5;241m8\u001b[39m))\n\u001b[1;32m     39\u001b[0m \u001b[38;5;66;03m# Custom function to format annotations with '%'\u001b[39;00m\n\u001b[1;32m     40\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mpercent_format\u001b[39m(val):\n",
      "\u001b[0;31mNameError\u001b[0m: name 'plt' is not defined"
     ]
    }
   ],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "# Assume overlap_counts is generated as per the previous code\n",
    "# We also need the total number of relevant genes for each PC\n",
    "total_genes_per_pc = {pc: len(genes) for pc, genes in relevant_genes_per_pc.items()}\n",
    "\n",
    "# Calculate the overlap percentage between each pair of PC components\n",
    "overlap_percentage = {}\n",
    "for pc_i in range(1, n_components_75 + 1):\n",
    "    pc_i_genes = set(relevant_genes_per_pc[f\"PC{pc_i}\"])\n",
    "    for pc_j in range(pc_i + 1, n_components_75 + 1):\n",
    "        pc_j_genes = set(relevant_genes_per_pc[f\"PC{pc_j}\"])\n",
    "        overlap_count = len(pc_i_genes & pc_j_genes)\n",
    "        min_total_genes = min(\n",
    "            total_genes_per_pc[f\"PC{pc_i}\"], total_genes_per_pc[f\"PC{pc_j}\"]\n",
    "        )\n",
    "        # Calculate the overlap percentage\n",
    "        overlap_percentage[f\"PC{pc_i} & PC{pc_j}\"] = int(\n",
    "            (overlap_count / min_total_genes) * 100\n",
    "        )  # Percentage as integer\n",
    "\n",
    "# Convert overlap_percentage dictionary to a DataFrame for heatmap visualization\n",
    "overlap_percentage_df = pd.DataFrame(\n",
    "    0,\n",
    "    index=[f\"PC{i}\" for i in range(1, n_components_75 + 1)],\n",
    "    columns=[f\"PC{i}\" for i in range(1, n_components_75 + 1)],\n",
    ")\n",
    "\n",
    "for pair, percent in overlap_percentage.items():\n",
    "    pc_i, pc_j = pair.split(\" & \")\n",
    "    overlap_percentage_df.loc[pc_i, pc_j] = percent\n",
    "    overlap_percentage_df.loc[pc_j, pc_i] = (\n",
    "        percent  # Mirror the percentage for both (i, j) and (j, i)\n",
    "    )\n",
    "\n",
    "# Set up the matplotlib figure\n",
    "plt.figure(figsize=(10, 8))\n",
    "\n",
    "\n",
    "# Custom function to format annotations with '%'\n",
    "def percent_format(val):\n",
    "    return f\"{int(val)}%\" if val > 0 else \"\"\n",
    "\n",
    "\n",
    "# Create a heatmap with annotated overlap percentages without decimals\n",
    "sns.heatmap(\n",
    "    overlap_percentage_df,\n",
    "    annot=overlap_percentage_df.applymap(percent_format),\n",
    "    fmt=\"\",\n",
    "    cmap=\"YlGnBu\",\n",
    "    square=True,\n",
    "    cbar_kws={\"label\": \"Overlap Percentage (%)\"},\n",
    ")\n",
    "plt.title(\"Percentage Overlap of Relevant Genes Between PCA Components\")\n",
    "plt.xlabel(\"PCA Components\")\n",
    "plt.ylabel(\"PCA Components\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "import streamlit as st\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn.decomposition import PCA\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Function to filter out repetitive genes (Soham's method)\n",
    "def filter_genes(data, threshold=0.01):\n",
    "    \"\"\"Filters genes that have little variance across samples.\"\"\"\n",
    "    variances = data.var(axis=0)\n",
    "    filtered_data = data.loc[:, variances > threshold]\n",
    "    return filtered_data\n",
    "\n",
    "# Sidebar to upload file\n",
    "st.sidebar.title(\"Upload Your Data\")\n",
    "uploaded_file = st.sidebar.file_uploader(\"Upload CSV File\", type=[\"csv\"])\n",
    "\n",
    "if uploaded_file is not None:\n",
    "    # Read the uploaded file\n",
    "    raw_data = pd.read_csv(uploaded_file)\n",
    "    st.write(\"Raw Data:\")\n",
    "    st.write(raw_data)\n",
    "\n",
    "    # Filter the genes based on variance\n",
    "    st.write(\"Filtered Data (After Removing Repetitive Genes):\")\n",
    "    filtered_data = filter_genes(raw_data)\n",
    "    st.write(filtered_data)\n",
    "\n",
    "    # PCA Analysis\n",
    "    pca = PCA()\n",
    "    pca.fit(filtered_data)\n",
    "    explained_variance = pca.explained_variance_ratio_\n",
    "\n",
    "    # Variance Explained Plot\n",
    "    st.write(\"Variance Explained by Each Number of Principal Components:\")\n",
    "    fig, ax = plt.subplots()\n",
    "    ax.plot(np.cumsum(explained_variance))\n",
    "    ax.set_xlabel('Number of Principal Components')\n",
    "    ax.set_ylabel('Cumulative Explained Variance')\n",
    "    st.pyplot(fig)\n",
    "\n",
    "    # Slider for PCA Visualization\n",
    "    st.write(\"Select Number of Principal Components for Visualization:\")\n",
    "    num_pcs = st.slider(\"Number of Principal Components\", 1, min(len(filtered_data.columns), len(filtered_data)), 2)\n",
    "\n",
    "    # Perform PCA with selected number of components\n",
    "    pca_selected = PCA(n_components=num_pcs)\n",
    "    pca_transformed = pca_selected.fit_transform(filtered_data)\n",
    "\n",
    "    # Display transformed PCA data\n",
    "    st.write(f\"PCA Transformed Data with {num_pcs} Components:\")\n",
    "    st.write(pd.DataFrame(pca_transformed, columns=[f\"PC{i+1}\" for i in range(num_pcs)]))\n",
    "\n",
    "    # PCA Scatter Plot\n",
    "    if num_pcs >= 2:\n",
    "        st.write(\"PCA Scatter Plot:\")\n",
    "        fig, ax = plt.subplots()\n",
    "        ax.scatter(pca_transformed[:, 0], pca_transformed[:, 1], alpha=0.5)\n",
    "        ax.set_xlabel(\"PC1\")\n",
    "        ax.set_ylabel(\"PC2\")\n",
    "        st.pyplot(fig)\n",
    "else:\n",
    "    st.write(\"Please upload a dataset to get started.\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
